app/models/template_invocation.rb

```diff
@@ -1,13 +1,24 @@
 class TemplateInvocation < ActiveRecord::Base
+  include ForemanTasks::Concerns::ActionSubject
 
   belongs_to :template, :class_name => 'JobTemplate', :foreign_key => 'template_id'
   belongs_to :job_invocation, :inverse_of => :template_invocations
   has_many :input_values, :class_name => 'TemplateInvocationInputValue', :dependent => :destroy
   has_one :targeting, :through => :job_invocation
 
-  include ForemanTasks::Concerns::ActionSubject
+  validates_associated :input_values
+  validate :provides_required_input_values
 
   def to_action_input
     { :id => id, :name => template.name }
   end
+
+  private
+
+  def provides_required_input_values
+    required_input_ids = self.template.template_inputs.where(:required => true).pluck(:id)
+    unless (required_input_ids - self.input_values.pluck(:template_input_id)).empty?
```

Comment:
Ahhh right, good call!

On Wed, Sep 9, 2015 at 3:50 AM, Marek Hulán <notifications@github.com>
wrote:

> In app/models/template_invocation.rb
> <https://github.com/theforeman/foreman_remote_execution/pull/17#discussion_r39015132>
> :
>
> >
> >    def to_action_input
> >      { :id => id, :name => template.name }
> >    end
> > +
> > +  private
> > +
> > +  def provides_required_input_values
> > +    required_input_ids = self.template.template_inputs.where(:required => true).pluck(:id)
> > +    unless (required_input_ids - self.input_values.pluck(:template_input_id)).empty?
>
> it's not only about testing, it didn't work for new records that were not
> saved yet since
>
> self.input_values # => return all including in-memory objectsself.input_values.pluck(...) # => returns only persisted objects
>
> which means in case of new record required_input_ids - always_empty_array
> which is not empty if there's any required input regardless of provided
> input values
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/theforeman/foreman_remote_execution/pull/17/files#r39015132>
> .
>